### PR TITLE
Corrected GQ -> LQ data acquisition in slices

### LINKED ISSUE #11452: Flaw in the GQ -> LQ data acquisition pagination logic
**Impact of the bug**
WMAgent

**Describe the bug**
We received a report of lack of pressure in the condor pool - via MM this morning - and the reason was that the agents were no longer acquiring work from the global workqueue.

**How to reproduce it**
Very specific scenario, but apparently caused by ~5k ACDCs injected into the system yesterday (likely a correlation of target sites and very high priority).

**Expected behavior**
When the agent is running the GQ to LQ data acquisition (where "data" here is a workqueue element), the expected behavior and logic is:
* it has to iterate over the GQE in slices/pages
* whenever the target number of elements is found (passed in the query string as `num_elem`), stop iterating in the CouchDB list function and return data to the client
* keep iterating through the elements until the whole data is exhausted, or until num_elem are found.

So, for this issue, we need to correct how pagination is done to ensure that the list `workRestrictions` will receive all the required pages from the view function `availableByPriority`, instead of passing over only the first page and returning results to the end user.

**Additional context and error message**
The agent uses `availableWork()` method to pull work from GQ to the LQ. A few options are defined:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WorkQueue/WorkQueueBackend.py#L446-L471

including the use of `skip` and `limit`, to act as a pagination over all the global workqueue elements and evaluate them into pages (slices).

This call gets mapped to something similar to (from the couchdb log):
```
GET /workqueue/_design/WorkQueue/_list/workRestrictions/availableByPriority?
include_docs=true&
descending=true&
resources=%7B%22T2_CH_CERN%22%3A+20975.0%2C+%22T2_CH_CERN_P5%22%3A+187....5.0%7D&
limit=1000&
num_elem=1000&
team=%22production%22&
skip=0
```

### PR DESCRIPTION
**Pull request by amaltaro**
Fixes #11452 

#### Status
not-tested

#### Description
Iterate through all the available elements, in slices of 2.5k docs, until the configured number of elements (200 atm) has been successfully acquired - or until all rows have been exhausted. This PR adds two new WorkQueueManager configuration parameters:
* `RowsPerSlice`: defining the slice size, i.e., how many rows/elements can be retrieved from workqueue within a single HTTP request;
* `MaxRowsPerCycle`: maximum number of rows/elements to be queried in workqueue. Currently 50k, meaning that if there are more than 50k elements in Available status, those won't be considered for data acquisition (they are, of course, ordered by priority though). 

An element "succcessfully" acquired means that:
* it passes all the work [restrictions](https://github.com/dmwm/WMCore/blob/master/src/couchapps/WorkQueue/lists/workRestrictions.js), especially matching the resources available
* and, during post processing in WorkQueueManager, that it has a priority higher than the workflows already in the queue. 

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
None

#### External dependencies / deployment changes
None

### CODE REVIEW

**Comment by amaltaro** [comment]
I had to test it in one of the production agents, to make sure it gets tested in the most meaningful scenario (many WQEs, diff priorities, busy agent/jobCounts, etc). I consider these tests successful, and from the logs below, we can see that it keeps iterating until `numElem` or `total_rows` is reached:
```
2023-01-26 21:01:18,448:140557188196096:INFO:WorkQueueBackend:Getting up to 200 available work from https://cmsweb.cern.ch/couchdb/workqueue
2023-01-26 21:01:18,448:140557188196096:INFO:WorkQueueBackend:  for team name: production
2023-01-26 21:01:18,448:140557188196096:INFO:WorkQueueBackend:  with excludeWorkflows count: 0
2023-01-26 21:01:18,448:140557188196096:INFO:WorkQueueBackend:  for thresholds: {'T2_US_Caltech': 252.5, ...}
2023-01-26 21:01:18,448:140557188196096:INFO:WorkQueueBackend:  with limit docs: 2500
2023-01-26 21:01:18,448:140557188196096:INFO:WorkQueueBackend:  and skip first docs: 0
2023-01-26 21:01:24,212:140557188196096:INFO:WorkQueueBackend:Retrieved 85 elements from workRestrictions list for: https://cmsweb.cern.ch/couchdb/workqueue
...Accepting a bunch of workflows ...
2023-01-26 21:01:24,221:140557188196096:INFO:WorkQueueBackend:  with limit docs: 2500
2023-01-26 21:01:24,221:140557188196096:INFO:WorkQueueBackend:  and skip first docs: 2500
2023-01-26 21:01:30,630:140557188196096:INFO:WorkQueueBackend:Retrieved 2500 elements from workRestrictions list for: https://cmsweb.cern.ch/couchdb/workqueue
2023-01-26 21:01:30,869:140557188196096:INFO:WorkQueueBackend:  with limit docs: 2500
2023-01-26 21:01:30,870:140557188196096:INFO:WorkQueueBackend:  and skip first docs: 5000
2023-01-26 21:01:38,527:140557188196096:INFO:WorkQueueBackend:Retrieved 2500 elements from workRestrictions list for: https://cmsweb.cern.ch/couchdb/workqueue
2023-01-26 21:01:38,784:140557188196096:INFO:WorkQueueBackend:  with limit docs: 2500
2023-01-26 21:01:38,784:140557188196096:INFO:WorkQueueBackend:  and skip first docs: 7500
2023-01-26 21:01:47,928:140557188196096:INFO:WorkQueueBackend:Retrieved 2341 elements from workRestrictions list for: https://cmsweb.cern.ch/couchdb/workqueue
2023-01-26 21:01:48,089:140557188196096:INFO:WorkQueueBackend:Accepting workflow: pdmvserv_task_SMP-RunIIFall18GS-00029__v1_T_221213_030638_1919, with prio: 70005, element id: 7dd9c4860f218f0fa688db6c865ee371, for site: T2_US_Florida
2023-01-26 21:01:48,124:140557188196096:INFO:WorkQueueBackend:And 42 elements passed location and siteJobCounts restrictions for: https://cmsweb.cern.ch/couchdb/workqueue
2023-01-26 21:01:48,293:140557188196096:INFO:WorkQueue:Assigned work to the child queue for:
2023-01-26 21:01:48,293:140557188196096:INFO:WorkQueue:    11 elements for: cmsunified_task_TAU-RunIISummer20UL16wmLHEGENAPV-00021__v1_T_230126_145940_3018
2023-01-26 21:01:48,293:140557188196096:INFO:WorkQueue:    1 elements for: cmsunified_task_HIG-RunIISummer20UL17wmLHEGEN-12349__v1_T_230126_153113_9874
2023-01-26 21:01:48,293:140557188196096:INFO:WorkQueue:    1 elements for: cmsunified_task_HIG-RunIISummer20UL16wmLHEGEN-13167__v1_T_230126_152920_2680
2023-01-26 21:01:48,293:140557188196096:INFO:WorkQueue:    1 elements for: cmsunified_task_B2G-RunIISummer20UL16wmLHEGENAPV-04172__v1_T_230126_150640_2473
2023-01-26 21:01:48,293:140557188196096:INFO:WorkQueue:    1 elements for: cmsunified_task_SMP-RunIISummer20UL16wmLHEGEN-00589__v1_T_230126_150421_1622
2023-01-26 21:01:48,293:140557188196096:INFO:WorkQueue:    1 elements for: cmsunified_task_SMP-RunIISummer20UL16wmLHEGEN-00586__v1_T_230126_155044_4584
2023-01-26 21:01:48,293:140557188196096:INFO:WorkQueue:    1 elements for: cmsunified_task_HIG-RunIISummer20UL16wmLHEGEN-13164__v1_T_230126_155906_5462
2023-01-26 21:01:48,293:140557188196096:INFO:WorkQueue:    24 elements for: cmsunified_task_TAU-RunIISummer20UL16wmLHEGEN-00015__v1_T_230126_155715_8597
2023-01-26 21:01:48,293:140557188196096:INFO:WorkQueue:    1 elements for: pdmvserv_task_SMP-RunIIFall18GS-00029__v1_T_221213_030638_1919
2023-01-26 21:01:48,294:140557188196096:INFO:WorkQueueManagerWorkPoller:Obtained 42 unit(s) of work
```

File: src/python/WMCore/WorkQueue/WorkQueueBackend.py [hunk 1]
```diff
@@ -458,28 +464,21 @@ def availableWork(self, thresholds, siteJobCounts, team=None,
...
         #  c) or, once "numElems" elements have been accepted
-        numSkip = 0
-        breakOut = False
-        while True:
-            if breakOut:
-                # then we have reached the maximum number of elements to be accepted
-                break
-            self.logger.info("  with limit docs: %s, and skip first %s docs", docsSliceSize, numSkip)
-            options['skip'] = numSkip
+        numSlices = ceil(numAvail / docsSliceSize)
```
Comment (vkuznet) [inline src/python/WMCore/WorkQueue/WorkQueueBackend.py:469]: Even though the code seems reasonable it is very hard to follow especially since someone needs to track the logic of slices. I rather prefer to get some code re-factoring (since changes are required anyway) and make it clear, e.g.
```
for sliceNum in range(numSlices):
      docs = self.getSliceOfDocs(sliceNum)
      wflows = self.getAvailableWorkflows(docs)
      siteJobCounts = self.getSiteJobCounts(wflows)
return elements, siteJobCounts
```
Such logic will be much easier to read and follow, where individual functions will perform relevant parts of current logic.

**Review by vkuznet — CHANGES_REQUESTED** [review]

File: src/python/WMCore/WorkQueue/WorkQueueBackend.py (hunk 1 above)
Comment (amaltaro) [reply src/python/WMCore/WorkQueue/WorkQueueBackend.py:469]: Yes, I've been considering to separate that block dealing with the elements and checking against siteJobCounts (basically verifying the priority of the work and having the final evaluation of whether the WQE can go through or not).
I will see what can be done tomorrow. But I don't plan to make any major refactorings now, especially because this will have to go through a complete change once we go to CouchDB 4.

Comment (amaltaro) [reply src/python/WMCore/WorkQueue/WorkQueueBackend.py:469]: Okay, the post-processing logic has been separated in its own method.

File: src/python/WMCore/WorkQueue/WorkQueueBackend.py [hunk 2]
```diff
@@ -458,87 +464,107 @@ def availableWork(self, thresholds, siteJobCounts, team=None,
...
-        return elements, siteJobCounts
+                         len(elems), self.queueUrl)
+        return elems
 
     def getActiveData(self):
         """Get data items we have work in the queue for"""
         data = self.db.loadView('WorkQueue', 'activeData', {'reduce': True, 'group': True})
-        return [{'dbs_url': x['key'][0],
-                 'name': x['key'][1]} for x in data.get('rows', [])]
+        return [{'dbs_url': x['key'][0], 'name': x['key'][1]} for x in data.get('rows', [])]
```
Comment (todor-ivanov) [inline src/python/WMCore/WorkQueue/WorkQueueBackend.py:567]: This line change does not contribute to the new logic. It is just a prettification of an existing line, which does not relate to the current issue and adds confusion. And I think I notice a pattern for breaking all of those similarly constructed lines down in the following  functions, obviously for better reading.

**Review by todor-ivanov — CHANGES_REQUESTED** [review]
Hi @amaltaro thanks for providing this PR.
I've left few comments inline. Please take a look.

File: src/python/WMCore/WorkQueue/WorkQueueBackend.py [hunk 3]
```diff
@@ -427,23 +428,28 @@ def availableWork(self, thresholds, siteJobCounts, team=None,
...
         self.logger.info("Getting up to %d available work from %s", numElems, self.queueUrl)
         self.logger.info("  for team name: %s", team)
-        self.logger.info("  with excludeWorkflows: %s", excludeWorkflows)
+        self.logger.info("  with excludeWorkflows count: %s", len(excludeWorkflows))
+        self.logger.debug("  with excludeWorkflows: %s", excludeWorkflows)
         self.logger.info("  for thresholds: %s", thresholds)
 
         # FIXME: magic numbers
-        docsSliceSize = 1000
+        docsSliceSize = 2500
```
[the remark this answers is not in the document]
Comment (amaltaro) [reply src/python/WMCore/WorkQueue/WorkQueueBackend.py:452]: This development is needed because there is a flaw in how this list is used, not because there is a "magic" number here. 
I have to check the code, passing variables in workqueue is usually not straight forward, due to the many layers of modules that it has to propagate through.

File: src/python/WMCore/WorkQueue/WorkQueueBackend.py [hunk 4]
```diff
@@ -458,87 +464,107 @@ def availableWork(self, thresholds, siteJobCounts, team=None,
...
-        breakOut = False
-        while True:
-            if breakOut:
-                # then we have reached the maximum number of elements to be accepted
-                break
-            self.logger.info("  with limit docs: %s, and skip first %s docs", docsSliceSize, numSkip)
-            options['skip'] = numSkip
+        numSlices = ceil(numAvail / docsSliceSize)
+        # limit it to the first 50k, in case there are tons of work in the system
+        numSlices = min(numSlices, int(50001 / docsSliceSize))
```
[the remark this answers is not in the document]
Comment (amaltaro) [reply src/python/WMCore/WorkQueue/WorkQueueBackend.py:471]: Now Unified is controlling the amount of work assigned, so it might be that we will never go beyond 100k available elements. From the CouchDB experts, skipping large amount of rows (whatever is the definition or large here) is discouraged, so yet another reason to having a max limit here. This is also a safe mechanism to ensure that the system remains responsive to whatever global workqueue scenario we get exposed to.

Let me check if I can easily make this one configurable as well.

File: src/python/WMCore/WorkQueue/WorkQueueBackend.py [hunk 5]
```diff
@@ -458,87 +464,107 @@ def availableWork(self, thresholds, siteJobCounts, team=None,
...
+        for sliceNum in range(numSlices):
+            # documents to skip as a function of the slice number
+            options['skip'] = sliceNum * docsSliceSize
+            self.logger.info("  with limit docs: %s", options['limit'])
+            self.logger.info("  and skip first docs: %s", options['skip'])
```
[the remark this answers is not in the document]
Comment (amaltaro) [reply src/python/WMCore/WorkQueue/WorkQueueBackend.py:476]: Yes, but I wanted to also give a cue on how the couchdb request is being constructed here (using skip and limit query strings). I will update this code though, I think it's indeed better

**Comment by amaltaro** [comment]
@todor-ivanov the 3rd commit in this PR has the changes that you requested in your review (I still need to test them though).

**Review by todor-ivanov — APPROVED** [review]
thanks @amaltaro 
looks good

**Review by vkuznet — APPROVED** [review]

**Comment by amaltaro** [comment]
No code changes applied in my last commit, I was just squashing all the 3 initial commits together.

**Comment by amaltaro** [comment]
@vkuznet Valentin, I was forgetting to use the new parameters when fetching data from the parent queue (they were applied only when fetching data from the local queue). I made the relevant changes and they are available in this 2nd commit, in case you want to have a look:
https://github.com/dmwm/WMCore/pull/11457/commits/87452dbbce4787d8cdfc2f2af607d1c69dcf81b0

**Comment by amaltaro** [comment]
Unit test failure comes from an unstable test. Merging it and will get it backported to the wmagent branch.
